### PR TITLE
feat: implement Google OAuth login

### LINKED ISSUE #137: [FEAT] Implement Authentication / Users / Organizations (including linking social on settings page)
### Description:

- Develop and integrate a comprehensive system for user authentication, organization management, and social account linking. This includes endpoints for user registration, login, organization creation, user-to-organization linking, and managing social accounts on the settings page.

**Acceptance Criteria**:
1. User Registration Endpoint:
- Endpoint: The endpoint /api/v1/auth/register should be accessible via a POST request.
- Request Body: The request body should include user details such as name, email, password, and optional organization details.
- Successful Registration: When valid details are provided, the endpoint returns a 201 Created status with a success message and user details.
- Email Already Exists: If an email is already registered, the endpoint returns a 409 Conflict status with an appropriate error message.
 2. User Login Endpoint:
- Endpoint: The endpoint /api/v1/auth/login should be accessible via a POST request.
- Request Body: The request body should include email and password.
- Successful Login: When valid credentials are provided, the endpoint returns a 200 OK status with a success message, JWT token, and user details.
- Invalid Credentials: If the email or password is incorrect, the endpoint returns a 401 Unauthorized status with an appropriate error message.

3. Organization Creation Endpoint:
- Endpoint: The endpoint /api/v1/organizations should be accessible via a POST request.
- Request Body: The request body should include the organization name and details.
- Successful Creation: When valid details are provided, the endpoint returns a 201 Created status with a success message and organization details.
- Organization Already Exists: If an organization with the same name already exists, the endpoint returns a 409 Conflict status with an appropriate error message.

4. Link User to Organization Endpoint:
- Endpoint: The endpoint /api/v1/organizations/{organization-id}/users/{user-id} should be accessible via a POST request.
- Request Body: The request body should include details for linking the user to the organization.
- Successful Linking: When valid organization-id and user-id are provided, the endpoint returns a 200 OK status with a success message.
- Invalid IDs: When an invalid organization-id or user-id format is provided, the endpoint returns a 400 Bad Request status with an appropriate error message.
- User or Organization Not Found: When an organization-id or user-id that does not exist is provided, the endpoint returns a 404 Not Found status with an appropriate error message.
5. Link Social Account Endpoint:
- Endpoint: The endpoint /api/v1/users/{user-id}/social-links should be accessible via a POST request.
- Request Body: The request body should include social account details such as provider and access token.
- Successful Linking: When valid details are provided, the endpoint returns a 200 OK status with a success message and linked social account details.
- Invalid Social Account Details: If the social account details are invalid, the endpoint returns a 400 Bad Request status with an appropriate error message.
- **Request and Response Examples:**

### User Registration Endpoint

Endpoint URL:

POST /api/v1/auth/register

Request Body:

```
json
{
  "name": "John Doe",
  "email": "john.doe@example.com",
  "password": "securePassword123",
  "organization": {
    "name": "Example Organization"
  }
}
```
Responses:

Success Response:

Code: 201 Created
Content:

```
json
{
  "message": "User registered successfully",
  "status_code": 201,
  "data": {
    "id": "user-id",
    "name": "John Doe",
    "email": "john.doe@example.com",
    "organization": {
      "id": "organization-id",
      "name": "Example Organization"
    },
    "created_at": "DateTime"
  }
}
```
Error Responses:

Code: 409 CONFLICT
Content:

```
json

{
  "status_code": 409,
  "message": "Email already exists"
}
```
User Login Endpoint

Endpoint URL:

POST /api/v1/auth/login

Request Body:

```
json

{
  "email": "john.doe@example.com",
  "password": "securePassword123"
}
```
Responses:

Success Response:

Code: 200 OK
Content:

```
json

{
  "message": "Login successful",
  "status_code": 200,
  "data": {
    "id": "user-id",
    "name": "John Doe",
    "email": "john.doe@example.com",
    "token": "jwt-token",
    "created_at": "DateTime"
  }
}
```
Error Responses:

Code: 401 UNAUTHORIZED
Content:

```
json
{
  "status_code": 401,
  "message": "Invalid credentials"
}
```
Organization Creation Endpoint

Endpoint URL:

POST /api/v1/organizations

Request Body:

```
json

{
  "name": "Example Organization"
}
```
Responses:

Success Response:

Code: 201 Created
Content:

```
json

{
  "message": "Organization created successfully",
  "status_code": 201,
  "data": {
    "id": "organization-id",
    "name": "Example Organization",
    "created_at": "DateTime"
  }
}
```
Error Responses:

Code: 409 CONFLICT
Content:

```
json

{
  "status_code": 409,
  "message": "Organization already exists"
}
```
**Link User to Organization Endpoint**

Endpoint URL:

POST /api/v1/organizations/{organization-id}/users/{user-id}

Request Body:
```

json
{
  "role": "admin"
}
```
Responses:

Success Response:

Code: 200 OK
Content:

```
json

{
  "message": "User linked to organization successfully",
  "status_code": 200,
  "data": {
    "user_id": "user-id",
    "organization_id": "organization-id",
    "role": "admin",
    "created_at": "DateTime"
  }
}

```
Error Responses:

Code: 400 BAD REQUEST
Content:

```
json

{
  "status_code": 400,
  "message": "Invalid organization ID or user ID format"
}
```
Code: 404 NOT FOUND
Content:

```
json

{
  "status_code": 404,
  "message": "Organization or user not found"
}
```
**Link Social Account Endpoint**

Endpoint URL:

POST /api/v1/users/{user-id}/social-links

Request Body:

```
json

{
  "provider": "google",
  "access_token": "socialAccessToken"
}
```
Responses:

Success Response:

Code: 200 OK
Content:

```
json

{
  "message": "Social account linked successfully",
  "status_code": 200,
  "data": {
    "id": "link-id",
    "user_id": "user-id",
    "provider": "google",
    "linked_at": "DateTime"
  }
}
```
Error Responses:

Code: 400 BAD REQUEST
Content:

```
json

{
  "status_code": 400,
  "message": "Invalid social account details"
}
```
Database Schema:
sql

```
Table users {
  id int [pk, increment]
  name varchar(255) [not null]
  email varchar(255) [not null, unique]
  password varchar(255) [not null]
  created_at timestamp [default: `current_timestamp`]
  updated_at timestamp [default: `current_timestamp`]
}

Table organizations {
  id int [pk, increment]
  name varchar(255) [not null]
  created_at timestamp [default: `current_timestamp`]
  updated_at timestamp [default: `current_timestamp`]
}

Table user_organizations {
  id int [pk, increment]
  user_id int [ref: > users.id]
  organization_id int [ref: > organizations.id]
  role varchar(255) [not null]
  created_at timestamp [default: `current_timestamp`]
  updated_at timestamp [default: `current_timestamp`]
}

Table social_links {
  id int [pk, increment]
  user_id int [ref: > users.id]
  provider varchar(255) [not null]
  access_token text [not null]
  created_at timestamp [default: `current_timestamp`]
  updated_at timestamp [default: `current_timestamp`]
}
```
### Purpose:

- The User Registration endpoint allows new users to register and optionally join an organization.
- The User Login endpoint handles user authentication and provides a JWT token for session management.
- The Organization Creation endpoint allows users to create new organizations.
- The Link User to Organization endpoint assigns a user to an organization with a specific role.
- The Link Social Account endpoint allows users to link their social accounts to their profile.

### Requirements:

- Implement endpoints for user registration, login, organization creation, user-to-organization linking, and social account linking.
- Return detailed information on created or linked entities.
- Handle scenarios where entities (users, organizations) do not exist or are invalid.

### PR DESCRIPTION
**Pull request by ZEEZKING**
# Description

This pull request refactors the Google login handling and updates the corresponding unit tests. It introduces an abstraction for Google token validation and ensures that the GoogleLoginCommandHandler works correctly with mocked dependencies.

**Closes #137**

# Changes proposed

## What were you told to do?

- Refactor the Google login handling in the GoogleLoginCommandHandler and update the unit tests to reflect the new implementation. The task involved abstracting the Google token validation logic and ensuring that the tests mock dependencies correctly.

## What did you do?

- Introduced IGoogleAuthService: Added an interface for abstracting Google token validation.
- Refactored GoogleLoginCommandHandler: Updated to use IGoogleAuthService for token validation and adjusted logic accordingly.
- Updated unit tests:
- Mocked IGoogleAuthService for testing purposes.
- Corrected return type setup for the AddAsync method in the user repository mock.
- Fixed setup for Google token validation and user repository interactions.

### CODE REVIEW

File: src/Hng.Application.Test/Features/UserManagement/GoogleLoginCommandHandlerTests.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Hng.Application.Test.Features.UserManagement
+{
+    public class GoogleLoginCommandHandlerTests
+    {
+        private readonly Mock<IRepository<User>> _userRepoMock;
+        private readonly Mock<ITokenService> _tokenServiceMock;
+        private readonly Mock<IMapper> _mapperMock;
```
Comment (chimaihueze) [inline src/Hng.Application.Test/Features/UserManagement/GoogleLoginCommandHandlerTests.cs:24]: do not mock Mapper. check tests written by others

File: src/Hng.Application/Features/UserManagement/Dtos/GoogleLoginRequestDto.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Hng.Application.Features.UserManagement.Dtos
+{
+    public class GoogleLoginRequestDto
```
Comment (chimaihueze) [inline src/Hng.Application/Features/UserManagement/Dtos/GoogleLoginRequestDto.cs:10]: does this dto map to any entity?

**Review by Godhanded — APPROVED** [review]

File: src/Hng.Web/Controllers/AuthenticationController.cs [hunk 3]
```diff
@@ -54,5 +54,27 @@ public async Task<ActionResult<SignUpResponse>> UserSignUp([FromBody] UserSignUp
...
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        public async Task<ActionResult<UserLoginResponseDto>> GoogleLogin([FromBody] GoogleLoginRequestDto googleLoginRequest)
+        {
+            var command = new GoogleLoginCommand(googleLoginRequest.IdToken);
+            var response = await _mediator.Send(command);
+
+            if (response == null || response.Data == null)
+            {
+                return Unauthorized(new
+                {
```
Comment (chimaihueze) [inline src/Hng.Web/Controllers/AuthenticationController.cs:69]: check the Application/Shared folder for dtos used for returning responses

**Review by chimaihueze — CHANGES_REQUESTED** [review]

File: src/Hng.Application/Features/UserManagement/Dtos/GoogleLoginRequestDto.cs (hunk 2 above)
Comment (ZEEZKING) [reply src/Hng.Application/Features/UserManagement/Dtos/GoogleLoginRequestDto.cs:10]: no

Comment (chimaihueze) [reply src/Hng.Application/Features/UserManagement/Dtos/GoogleLoginRequestDto.cs:10]: > no

ok. handle other feedbacks, and I will merge.

**Review by chimaihueze — APPROVED** [review]

**Pull request merged**